File: ios/entitlements.rb
require('cfpropertylist')

require_relative('pod_helpers')

DEFAULT_IOS_ENTITLEMENTS = {
  'keychain-access-groups' => [
    '$(AppIdentifierPrefix)com.microsoft.adalcache',
  ],
}.freeze

DEFAULT_MACOS_ENTITLEMENTS = {
  'com.apple.security.app-sandbox' => true,
  'com.apple.security.files.user-selected.read-only' => true,
  'com.apple.security.network.client' => true,
}.freeze

def generate_entitlements!(project_root, target_platform, destination)
  user_entitlements = platform_config('codeSignEntitlements', project_root, target_platform)
  # If `codeSignEntitlements` is a string, set `CODE_SIGN_ENTITLEMENTS` instead
  return if user_entitlements.is_a? String

  entitlements = target_platform == :macos ? DEFAULT_MACOS_ENTITLEMENTS : DEFAULT_IOS_ENTITLEMENTS

  plist = CFPropertyList::List.new
  plist.value = CFPropertyList.guess(entitlements.merge(user_entitlements || {}))
  plist.save(File.join(destination, 'App.entitlements'),
             CFPropertyList::List::FORMAT_XML,
             { :formatted => true })
end

File: ios/entitlements.mjs
// @ts-check
import * as nodefs from "node:fs";
import * as path from "node:path";
import { isObject, toPlist } from "./utils.mjs";

/** @import { ApplePlatform, JSONObject } from "../scripts/types.js"; */

const DEFAULT_IOS_ENTITLEMENTS = {
  "keychain-access-groups": ["$(AppIdentifierPrefix)com.microsoft.adalcache"],
};

const DEFAULT_MACOS_ENTITLEMENTS = {
  "com.apple.security.app-sandbox": true,
  "com.apple.security.files.user-selected.read-only": true,
  "com.apple.security.network.client": true,
};

/**
 * @param {JSONObject} appConfig
 * @param {ApplePlatform} targetPlatform
 * @returns {string | JSONObject | undefined}
 */
function getCodeSignEntitlements(appConfig, targetPlatform) {
  const platformConfig = appConfig[targetPlatform];
  if (!isObject(platformConfig)) {
    return;
  }

  const codeSignEntitlements = platformConfig["codeSignEntitlements"];
  if (
    typeof codeSignEntitlements !== "string" &&
    !isObject(codeSignEntitlements)
  ) {
    return;
  }

  return codeSignEntitlements;
}

/**
 * @param {JSONObject} appConfig
 * @param {ApplePlatform} targetPlatform
 * @param {string} destination
 * @returns {void}
 */
export function generateEntitlements(
  appConfig,
  targetPlatform,
  destination,
  fs = nodefs
) {
  // If `codeSignEntitlements` is a string, set `CODE_SIGN_ENTITLEMENTS` instead
  const userEntitlements = getCodeSignEntitlements(appConfig, targetPlatform);
  if (typeof userEntitlements === "string") {
    return;
  }

  const filename = "App.entitlements";
  const entitlements = toPlist(
    {
      ...(targetPlatform === "macos"
        ? DEFAULT_MACOS_ENTITLEMENTS
        : DEFAULT_IOS_ENTITLEMENTS),
      ...userEntitlements,
    },
    filename
  );

  fs.writeFileSync(path.join(destination, filename), entitlements);
}
